### PR TITLE
add coingecko price deviation event

### PR DESCRIPTION
**Pull request by cctdaniel**
- still fixing ratelimit

### CODE REVIEW

File: tests/test_price_deviation_coingecko.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+from pyth_observer.events import PriceDeviation, PriceDeviationCoinGecko
+from pythclient.pythaccounts import PythPriceInfo, PythPriceStatus
+from pyth_observer.prices import Price
+
+
+class MockPythProductAccount:
```
Comment (tompntn) [inline tests/test_price_deviation_coingecko.py:6]: Have you seen [unittest.mock](https://docs.python.org/3/library/unittest.mock.html)? I've found this super useful for easily mocking python objects with minimal effort/boilerplate.

Comment (cctdaniel) [reply tests/test_price_deviation_coingecko.py:6]: yup I used that for https://github.com/pyth-network/pyth-client-py -- was trying to follow the existing tests which was using a mock class, happy to migrate all to unittest in a follow up PR

File: pyth_observer/events.py [hunk 2]
```diff
@@ -383,3 +386,35 @@ def get_event_details(self) -> Tuple[str, List[str]]:
...
+    """
+    This alert is supposed to fire when a price feed deviates from CoinGecko price feed by a specified threshold.
+    """
+    error_code: str = "price-deviation-coingecko"
+    threshold = int(os.environ.get('PYTH_OBSERVER_PRICE_DEVIATION_COINGECKO', 5))
+
+    def is_valid(self) -> bool:
+        pyth_price = self.price_account.aggregate_price_info.price
+        if self.coingecko_price is None or pyth_price == 0:
+            # TODO: add another alert that validates whether the aggregate price is close to the truth
```
Comment (jayantk) [inline pyth_observer/events.py:401]: i think this todo is stale

File: pyth_observer/coingecko.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    with open(f'{ROOT_DIR}/coingecko_mapping.json') as json_file:
+        data = json.load(json_file)
+
+    return data
+
+
+mapping = get_coingecko_symbol_to_id_mapping()
+
+
+def get_coingecko_prices(symbols):
```
Comment (jayantk) [inline pyth_observer/coingecko.py:22]: this function exposes the coingecko ids. If you re-mapped the symbols -> prices inside this function, then the caller wouldn't have to call get_coingecko_api_id

File: observer.py [hunk 4]
```diff
@@ -65,16 +66,21 @@ async def main(args):
                 asyncio.sleep(0.4)
                 continue
 
+            crypto_products = set([p.attrs["base"] for p in products if p.attrs["asset_type"] == 'Crypto'])
+            coingecko_prices = get_coingecko_prices(crypto_products)
```
Comment (jayantk) [inline observer.py:70]: i think this price update should run in a separate asyncio loop and cache the results in a variable. this loop can then read the most recent prices from that variable. You definitely don't want to couple together the 0.4s alerting times with the 30s coingecko rate limit, nor do you want to block the alerts while waiting for the coingecko api to return.

File: pyth_observer/coingecko.py (hunk 3 above)
Comment (cctdaniel) [reply pyth_observer/coingecko.py:22]: just out of curiosity, what is the concern for exposing the coingecko ids?

Comment (jayantk) [reply pyth_observer/coingecko.py:22]: It's a software engineering thing. There's no reason for the caller to know that these prices even come from coingecko -- they just want a method like `get_prices` that downloads price data for pyth symbols. That method could call coingecko or some other API. If the interface hides how exactly the prices are obtained, it's easy to switch to a different API later. If the interface exposes coingecko-specific stuff, then it's not so easy.

Comment (cctdaniel) [reply pyth_observer/coingecko.py:22]: got it, thank you for the detailed explanation 🙂

**Review by jayantk — APPROVED** [review]

**Pull request merged**